libraries/AP_InertialSensor/AP_InertialSensor_Logging.cpp

```diff
@@ -0,0 +1,150 @@
+        instance  : instance,
+        sample_us : sample_us?sample_us:now,
+        AccX      : accel.x,
+        AccY      : accel.y,
+        AccZ      : accel.z
+    };
+    AP::logger().WriteBlock(&pkt, sizeof(pkt));
+}
+
+void AP_InertialSensor_Backend::Write_GYR(const uint8_t instance, const uint64_t sample_us, const Vector3f &gyro) const 
+{
+    uint64_t now = AP_HAL::micros64();
+    const struct log_GYR pkt{
+        LOG_PACKET_HEADER_INIT(LOG_GYR_MSG),
+        time_us   : now,
+        instance  : instance,
+        sample_us : sample_us?sample_us:now,
+        GyrX      : gyro.x,
+        GyrY      : gyro.y,
+        GyrZ      : gyro.z
+    };
+    AP::logger().WriteBlock(&pkt, sizeof(pkt));
+}
+
+// Write IMU data packet: raw accel/gyro data
+void AP_InertialSensor::Write_IMU_instance(const uint64_t time_us, const uint8_t imu_instance) const
+{
+    const Vector3f &gyro = get_gyro(imu_instance);
+    const Vector3f &accel = get_accel(imu_instance);
+    const struct log_IMU pkt{
+        LOG_PACKET_HEADER_INIT(LOG_IMU_MSG),
+        time_us : time_us,
+        instance: imu_instance,
+        gyro_x  : gyro.x,
+        gyro_y  : gyro.y,
+        gyro_z  : gyro.z,
+        accel_x : accel.x,
+        accel_y : accel.y,
+        accel_z : accel.z,
+        gyro_error  : get_gyro_error_count(imu_instance),
+        accel_error : get_accel_error_count(imu_instance),
+        temperature : get_temperature(imu_instance),
+        gyro_health : (uint8_t)get_gyro_health(imu_instance),
+        accel_health : (uint8_t)get_accel_health(imu_instance),
+        gyro_rate : get_gyro_rate_hz(imu_instance),
+        accel_rate : get_accel_rate_hz(imu_instance),
+    };
+    AP::logger().WriteBlock(&pkt, sizeof(pkt));
+}
+
+// Write IMU data packet for all instances
+void AP_InertialSensor::Write_IMU() const
+{
+    const uint64_t time_us = AP_HAL::micros64();
+
+    uint8_t n = MAX(get_accel_count(), get_gyro_count());
+    for (uint8_t i=0; i<n; i++) {
+        Write_IMU_instance(time_us, i);
+    }
+}
+
+// Write VIBE data packet for all instances
+void AP_InertialSensor::Write_Vibration() const
+{
+    const uint64_t time_us = AP_HAL::micros64();
+    for (uint8_t i = 0; i < INS_MAX_INSTANCES; i++) {
+        if (!use_accel(i)) {
+            continue;
+        }
+
+        const Vector3f vibration = get_vibration_levels(i);
+        const struct log_Vibe pkt{
+            LOG_PACKET_HEADER_INIT(LOG_VIBE_MSG),
+            time_us     : time_us,
+            imu         : i,
+            vibe_x      : vibration.x,
+            vibe_y      : vibration.y,
+            vibe_z      : vibration.z,
+            clipping    : get_accel_clip_count(i)
+        };
+        AP::logger().WriteBlock(&pkt, sizeof(pkt));
+    }
+}
+
+// Write information about a series of IMU readings to log:
+bool AP_InertialSensor::BatchSampler::Write_ISBH(const float sample_rate_hz) const
+{
+    // if (_next_backend == 0) {
+    //     return false;
+    // }
```

Comment:
You've lost this rather important check in your new function.  That's a method-on-nullptr bug, which ++ungood.

Reply:
Hear you on the `Write_GYR` and `Write_ACC`! I will submit those separately. 

On the nullptr check. I noticed the `AP_Logger::WriteBlock` neglects that as it presumes that callers use `AP_Logger::should_log`. 

Would it make sense to add the following here?https://github.com/ArduPilot/ardupilot/blob/1d5b62c3d8bf95598b927c88b4a271df1f3fb11b/libraries/AP_InertialSensor/BatchSampler.cpp#L194

```cpp
#define MASK_LOG_ANY 0xFFFF    
if (!logger->should_log(MASK_LOG_ANY)) {  return false; }
```

This would be done per your previous comment separately?